<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Save/load SpatialFeatureExperiment to/from file • alabaster.sfe</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Save/load SpatialFeatureExperiment to/from file">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">alabaster.sfe</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Overview.html">Save/load SpatialFeatureExperiment to/from file</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Save/load SpatialFeatureExperiment to/from file</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
            
            <h4 data-toc-skip class="date">2024-12-04</h4>
      

      <div class="d-none name"><code>Overview.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>The <code>SpatialFeatureExperiment</code> (SFE) class extends
<code>SpatialExperiment</code> to foreground the spatial aspect of the
data by incorporating Simple Feature geometries and geometric operations
on the whole object like a multi-layered map. It also introduces new
image classes to support the geospatial raster tradition
(<code>terra</code> package) and the image processing package
(<code>EBImage</code> package).</p>
<p>As spatial -omics data is growing in size, some components of the SFE
object are not loaded into memory unless necessary. This includes gene
count matrix and assays that are <code>DelayedArrays</code> and
BioFormats images only whose metadata is loaded into memory, as well as
transcript spot geometries for smFISH-based technologies in some cases.
We plan to expand the out of memory functionalities in the near future.
As a result, when the SFE object is save to RDS and shared with others,
often the out of memory components break because the original file is
unavailable or is tied to some absolute path.</p>
<p>A goal of <code>alabaster.sfe</code> is to make SFE objects more
portable for sharing, and to avoid recomputing affine transformations
and spatial statistics after reading the data from the original output
such as from Space Ranger. It is a contribution to the
<code>alabaster</code> suite of packages from <a href="https://github.com/ArtifactDB" class="external-link">ArtifactDB</a>. Another goal is
interoperability with Python and other languages as in the goal of
ArtifactDB.</p>
</div>
<div class="section level2">
<h2 id="xenium-demo">Xenium demo<a class="anchor" aria-label="anchor" href="#xenium-demo"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://pachterlab.github.io/alabaster.sfe/">alabaster.sfe</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SpatialFeatureExperiment</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'SpatialFeatureExperiment'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     scale</span></span>
<span><span class="co">#&gt; Loading required package: alabaster.base</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SummarizedExperiment</span></span>
<span><span class="co">#&gt; Loading required package: MatrixGenerics</span></span>
<span><span class="co">#&gt; Loading required package: matrixStats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'matrixStats'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:alabaster.base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; Loading required package: generics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'generics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     as.difftime, as.factor, as.ordered, intersect, is.element, setdiff,</span></span>
<span><span class="co">#&gt;     setequal, union</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:SpatialFeatureExperiment':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     saveRDS</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, is.unsorted, lapply, Map, mapply, match, mget,</span></span>
<span><span class="co">#&gt;     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span></span>
<span><span class="co">#&gt;     rbind, Reduce, rownames, sapply, saveRDS, table, tapply, unique,</span></span>
<span><span class="co">#&gt;     unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:utils':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     findMatches</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: IRanges</span></span>
<span><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span><span class="co">#&gt; Loading required package: Biobase</span></span>
<span><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biobase'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rowMedians</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing, rowMedians</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:alabaster.base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: scuttle</span></span>
<span><span class="co">#&gt; Loading required package: ggplot2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'ggplot2'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:SpatialFeatureExperiment':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     unit</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-spatial.github.io/sf/" class="external-link">sf</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Linking to GEOS 3.10.2, GDAL 3.4.1, PROJ 8.2.1; sf_use_s2() is TRUE</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://fs.r-lib.org" class="external-link">fs</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'fs'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:BiocGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     path</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/r-spatial/spdep/" class="external-link">spdep</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: spData</span></span>
<span><span class="co">#&gt; To access larger datasets in this package, install the spDataLarge</span></span>
<span><span class="co">#&gt; package with: `install.packages('spDataLarge',</span></span>
<span><span class="co">#&gt; repos='https://nowosad.github.io/drat/', type='source')`</span></span>
<span><span class="fu"><a href="https://r-spatial.github.io/spdep/reference/set.spChkOption.html" class="external-link">set.SubgraphOption</a></span><span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Here we show a small example on a subset of Xenium data, doing some
basic spatial analyses.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">x2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/XeniumOutput.html" class="external-link">XeniumOutput</a></span><span class="op">(</span><span class="st">"v2"</span>, file_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fp</span>, <span class="st">"xenium2"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; The downloaded files are in /tmp/RtmpbPifJk/file269aa5e1f9f/xenium2/xenium2</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readXenium.html" class="external-link">readXenium</a></span><span class="op">(</span><span class="va">x2</span>, add_molecules <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Must use gene symbols as row names when adding transcript spots.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Cell segmentations are found in `.parquet` file(s)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading cell and nucleus segmentations</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Making MULTIPOLYGON nuclei geometries</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Making POLYGON cell geometries</span></span>
<span><span class="co">#&gt; Sanity checks on cell segmentation polygons:</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; ..found 132 cells with (nested) polygon lists</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; ..applying filtering</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Checking polygon validity</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Saving geometries to parquet files</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading cell metadata -&gt; `cells.parquet`</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading h5 gene count matrix</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; filtering cellSeg geometries to match 6272 cells with counts &gt; 0</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; filtering nucSeg geometries to match 6158 cells with counts &gt; 0</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading transcript coordinates</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Converting transcript spots to geometry</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Writing reformatted transcript spots to disk</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Total of 116 features/genes with no transcript detected or `min_phred` &lt; 20 are removed from SFE object</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; To keep all features -&gt; set `min_phred = NULL`</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 6272 rows and 9 columns</span></span>
<span><span class="co">#&gt;            transcript_counts control_probe_counts control_codeword_counts</span></span>
<span><span class="co">#&gt;                    &lt;integer&gt;            &lt;integer&gt;               &lt;integer&gt;</span></span>
<span><span class="co">#&gt; abclkehb-1                26                    0                       0</span></span>
<span><span class="co">#&gt; abcnopgp-1                35                    0                       0</span></span>
<span><span class="co">#&gt; abcobdon-1                39                    0                       0</span></span>
<span><span class="co">#&gt; abcohgbl-1                45                    0                       0</span></span>
<span><span class="co">#&gt; abcoochm-1                61                    0                       0</span></span>
<span><span class="co">#&gt; ...                      ...                  ...                     ...</span></span>
<span><span class="co">#&gt; odmfjplg-1                12                    0                       0</span></span>
<span><span class="co">#&gt; odmfpjop-1                 3                    0                       0</span></span>
<span><span class="co">#&gt; odmglico-1                40                    0                       0</span></span>
<span><span class="co">#&gt; odmgoega-1                25                    0                       0</span></span>
<span><span class="co">#&gt; odmgojlc-1                13                    0                       0</span></span>
<span><span class="co">#&gt;            unassigned_codeword_counts deprecated_codeword_counts total_counts</span></span>
<span><span class="co">#&gt;                             &lt;integer&gt;                  &lt;integer&gt;    &lt;integer&gt;</span></span>
<span><span class="co">#&gt; abclkehb-1                          0                          0           26</span></span>
<span><span class="co">#&gt; abcnopgp-1                          0                          0           35</span></span>
<span><span class="co">#&gt; abcobdon-1                          0                          0           39</span></span>
<span><span class="co">#&gt; abcohgbl-1                          0                          0           45</span></span>
<span><span class="co">#&gt; abcoochm-1                          0                          0           61</span></span>
<span><span class="co">#&gt; ...                               ...                        ...          ...</span></span>
<span><span class="co">#&gt; odmfjplg-1                          0                          0           12</span></span>
<span><span class="co">#&gt; odmfpjop-1                          0                          0            3</span></span>
<span><span class="co">#&gt; odmglico-1                          0                          0           40</span></span>
<span><span class="co">#&gt; odmgoega-1                          0                          0           25</span></span>
<span><span class="co">#&gt; odmgojlc-1                          0                          0           13</span></span>
<span><span class="co">#&gt;            cell_area nucleus_area   sample_id</span></span>
<span><span class="co">#&gt;            &lt;numeric&gt;    &lt;numeric&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt; abclkehb-1   39.6923      18.9656    sample01</span></span>
<span><span class="co">#&gt; abcnopgp-1   60.7352      29.5322    sample01</span></span>
<span><span class="co">#&gt; abcobdon-1   55.4519      20.9977    sample01</span></span>
<span><span class="co">#&gt; abcohgbl-1   79.9266      24.4295    sample01</span></span>
<span><span class="co">#&gt; abcoochm-1   70.6244      34.1833    sample01</span></span>
<span><span class="co">#&gt; ...              ...          ...         ...</span></span>
<span><span class="co">#&gt; odmfjplg-1  63.03813     29.35156    sample01</span></span>
<span><span class="co">#&gt; odmfpjop-1   9.16672      9.16672    sample01</span></span>
<span><span class="co">#&gt; odmglico-1  82.54563     16.07563    sample01</span></span>
<span><span class="co">#&gt; odmgoega-1 112.84547      5.73484    sample01</span></span>
<span><span class="co">#&gt; odmgojlc-1  47.00766     12.19219    sample01</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"knn5"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe</span>, method <span class="op">=</span> <span class="st">"knearneigh"</span>, k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe</span>, size_factors <span class="op">=</span> <span class="va">sfe</span><span class="op">$</span><span class="va">cell_area</span><span class="op">)</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">runMoransI</a></span><span class="op">(</span><span class="va">sfe</span>, colGraphName <span class="op">=</span> <span class="st">"knn5"</span><span class="op">)</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">colDataMoransI</a></span><span class="op">(</span><span class="va">sfe</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"transcript_counts"</span>, <span class="st">"cell_area"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">top_moran</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">moran_sample01</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateUnivariate.html" class="external-link">runUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, type <span class="op">=</span> <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="va">top_moran</span><span class="op">)</span></span></code></pre></div>
<p>Global spatial statistics for gene expression are stored in
<code>rowData</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 398 rows and 5 columns</span></span>
<span><span class="co">#&gt;                                             ID                 Symbol</span></span>
<span><span class="co">#&gt;                                    &lt;character&gt;            &lt;character&gt;</span></span>
<span><span class="co">#&gt; ABCC11                         ENSG00000121270                 ABCC11</span></span>
<span><span class="co">#&gt; ACE2                           ENSG00000130234                   ACE2</span></span>
<span><span class="co">#&gt; ACKR1                          ENSG00000213088                  ACKR1</span></span>
<span><span class="co">#&gt; ACTA2                          ENSG00000107796                  ACTA2</span></span>
<span><span class="co">#&gt; ACTG2                          ENSG00000163017                  ACTG2</span></span>
<span><span class="co">#&gt; ...                                        ...                    ...</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0461 UnassignedCodeword_0.. UnassignedCodeword_0..</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0469 UnassignedCodeword_0.. UnassignedCodeword_0..</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0479 UnassignedCodeword_0.. UnassignedCodeword_0..</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0488 UnassignedCodeword_0.. UnassignedCodeword_0..</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0497 UnassignedCodeword_0.. UnassignedCodeword_0..</span></span>
<span><span class="co">#&gt;                                        Type moran_sample01 K_sample01</span></span>
<span><span class="co">#&gt;                                 &lt;character&gt;      &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ABCC11                      Gene Expression    -0.00108527 1965.06388</span></span>
<span><span class="co">#&gt; ACE2                        Gene Expression     0.35448582   25.74963</span></span>
<span><span class="co">#&gt; ACKR1                       Gene Expression     0.13760601  140.94728</span></span>
<span><span class="co">#&gt; ACTA2                       Gene Expression     0.12094896   11.71618</span></span>
<span><span class="co">#&gt; ACTG2                       Gene Expression     0.27008958    8.84807</span></span>
<span><span class="co">#&gt; ...                                     ...            ...        ...</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0461 Unassigned Codeword   -0.000127571       6270</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0469 Unassigned Codeword   -0.000127571       6270</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0479 Unassigned Codeword   -0.000159464       6270</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0488 Unassigned Codeword   -0.000287036       6270</span></span>
<span><span class="co">#&gt; UnassignedCodeword_0497 Unassigned Codeword   -0.000191357       6270</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, features <span class="op">=</span> <span class="va">top_moran</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Overview_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>Global spatial statistics for <code>colData</code> can be accessed
with <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colFeatureData.html" class="external-link">colFeatureData()</a></code></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colFeatureData.html" class="external-link">colFeatureData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 10 rows and 2 columns</span></span>
<span><span class="co">#&gt;                            moran_sample01 K_sample01</span></span>
<span><span class="co">#&gt;                                 &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; transcript_counts                0.385497    24.5122</span></span>
<span><span class="co">#&gt; control_probe_counts                   NA         NA</span></span>
<span><span class="co">#&gt; control_codeword_counts                NA         NA</span></span>
<span><span class="co">#&gt; unassigned_codeword_counts             NA         NA</span></span>
<span><span class="co">#&gt; deprecated_codeword_counts             NA         NA</span></span>
<span><span class="co">#&gt; total_counts                           NA         NA</span></span>
<span><span class="co">#&gt; cell_area                        0.243229    17.2938</span></span>
<span><span class="co">#&gt; nucleus_area                           NA         NA</span></span>
<span><span class="co">#&gt; sample_id                              NA         NA</span></span>
<span><span class="co">#&gt; sizeFactor                             NA         NA</span></span></code></pre></div>
<p>Local spatial statistics are stored in the field
<code>localResults</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotLocalResult.html" class="external-link">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>, name <span class="op">=</span> <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="va">top_moran</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, diverge_center <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="Overview_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>So the SFE object can get quite a bit more complicated than
<code>SpatialExperiment</code> as the results from spatial analyses are
stored within the object to ease bookkeeping and plotting. The spatial
analysis results will be save to disk with
<code>alabaster.sfe</code>.</p>
<p>Here the image is an OME-TIFF and only the metadata is in memory.
Only the highest resolution necessary is loaded into memory for
plotting, and only the part being plotted is loaded.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotImage.html" class="external-link">plotImage</a></span><span class="op">(</span><span class="va">sfe</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, channel <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">1</span>, normalize_channels <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Overview_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>In addition, we can do an affine transform on the object. Say to
rotate it.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu">SpatialFeatureExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/SFE-transform.html" class="external-link">rotate</a></span><span class="op">(</span><span class="va">sfe</span>, degrees <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotGeometry.html" class="external-link">plotGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="st">"nucSeg"</span>, fill <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>             image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>,</span>
<span>             channel <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">1</span>, normalize_channels <span class="op">=</span> <span class="cn">TRUE</span>, dark <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Overview_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>Now we save this object to disk.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fsave</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fp</span>, <span class="st">"sfe_save"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/alabaster.base/man/saveObject.html" class="external-link">saveObject</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">fsave</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Saving SpatialExperiment</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Saving colgeometries</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Saving rowgeometries</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Saving localResults</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Saving spatial graphs</span></span></code></pre></div>
<p>The <code>SpatialExperiment</code> method is called first, which then
calls new methods to save images implemented in
<code>alabaster.sfe</code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_tree.html" class="external-link">dir_tree</a></span><span class="op">(</span><span class="va">fsave</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB; font-weight: bold;">/tmp/RtmpbPifJk/file269aa5e1f9f/sfe_save</span></span></span>
<span><span class="co">#&gt; ├── OBJECT</span></span>
<span><span class="co">#&gt; ├── <span style="color: #0000BB; font-weight: bold;">assays</span></span></span>
<span><span class="co">#&gt; │   ├── <span style="color: #0000BB; font-weight: bold;">0</span></span></span>
<span><span class="co">#&gt; │   │   ├── OBJECT</span></span>
<span><span class="co">#&gt; │   │   └── matrix.h5</span></span>
<span><span class="co">#&gt; │   ├── <span style="color: #0000BB; font-weight: bold;">1</span></span></span>
<span><span class="co">#&gt; │   │   ├── OBJECT</span></span>
<span><span class="co">#&gt; │   │   └── matrix.h5</span></span>
<span><span class="co">#&gt; │   └── names.json</span></span>
<span><span class="co">#&gt; ├── <span style="color: #0000BB; font-weight: bold;">colgeometries</span></span></span>
<span><span class="co">#&gt; │   ├── <span style="color: #0000BB; font-weight: bold;">0</span></span></span>
<span><span class="co">#&gt; │   │   ├── OBJECT</span></span>
<span><span class="co">#&gt; │   │   └── map.parquet</span></span>
<span><span class="co">#&gt; │   ├── <span style="color: #0000BB; font-weight: bold;">1</span></span></span>
<span><span class="co">#&gt; │   │   ├── OBJECT</span></span>
<span><span class="co">#&gt; │   │   └── map.parquet</span></span>
<span><span class="co">#&gt; │   ├── <span style="color: #0000BB; font-weight: bold;">2</span></span></span>
<span><span class="co">#&gt; │   │   ├── OBJECT</span></span>
<span><span class="co">#&gt; │   │   └── map.parquet</span></span>
<span><span class="co">#&gt; │   └── names.json</span></span>
<span><span class="co">#&gt; ├── <span style="color: #0000BB; font-weight: bold;">column_data</span></span></span>
<span><span class="co">#&gt; │   ├── OBJECT</span></span>
<span><span class="co">#&gt; │   ├── basic_columns.h5</span></span>
<span><span class="co">#&gt; │   ├── <span style="color: #0000BB; font-weight: bold;">column_annotations</span></span></span>
<span><span class="co">#&gt; │   │   ├── OBJECT</span></span>
<span><span class="co">#&gt; │   │   └── basic_columns.h5</span></span>
<span><span class="co">#&gt; │   └── <span style="color: #0000BB; font-weight: bold;">other_annotations</span></span></span>
<span><span class="co">#&gt; │       ├── OBJECT</span></span>
<span><span class="co">#&gt; │       └── <span style="color: #BB0000; font-weight: bold;">list_contents.json.gz</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #0000BB; font-weight: bold;">coordinates</span></span></span>
<span><span class="co">#&gt; │   ├── OBJECT</span></span>
<span><span class="co">#&gt; │   └── array.h5</span></span>
<span><span class="co">#&gt; ├── <span style="color: #0000BB; font-weight: bold;">images</span></span></span>
<span><span class="co">#&gt; │   ├── <span style="color: #0000BB; font-weight: bold;">0</span></span></span>
<span><span class="co">#&gt; │   │   ├── OBJECT</span></span>
<span><span class="co">#&gt; │   │   └── <span style="color: #0000BB; font-weight: bold;">image</span></span></span>
<span><span class="co">#&gt; │   │       ├── <span style="color: #BB00BB; font-weight: bold;">morphology_focus_0000.ome.tif</span></span></span>
<span><span class="co">#&gt; │   │       ├── <span style="color: #BB00BB; font-weight: bold;">morphology_focus_0001.ome.tif</span></span></span>
<span><span class="co">#&gt; │   │       ├── <span style="color: #BB00BB; font-weight: bold;">morphology_focus_0002.ome.tif</span></span></span>
<span><span class="co">#&gt; │   │       └── <span style="color: #BB00BB; font-weight: bold;">morphology_focus_0003.ome.tif</span></span></span>
<span><span class="co">#&gt; │   └── mapping.h5</span></span>
<span><span class="co">#&gt; ├── <span style="color: #0000BB; font-weight: bold;">local_results</span></span></span>
<span><span class="co">#&gt; │   ├── OBJECT</span></span>
<span><span class="co">#&gt; │   ├── basic_columns.h5</span></span>
<span><span class="co">#&gt; │   ├── <span style="color: #0000BB; font-weight: bold;">other_annotations</span></span></span>
<span><span class="co">#&gt; │   │   ├── OBJECT</span></span>
<span><span class="co">#&gt; │   │   └── <span style="color: #BB0000; font-weight: bold;">list_contents.json.gz</span></span></span>
<span><span class="co">#&gt; │   └── <span style="color: #0000BB; font-weight: bold;">other_columns</span></span></span>
<span><span class="co">#&gt; │       └── <span style="color: #0000BB; font-weight: bold;">0</span></span></span>
<span><span class="co">#&gt; │           ├── OBJECT</span></span>
<span><span class="co">#&gt; │           ├── basic_columns.h5</span></span>
<span><span class="co">#&gt; │           └── <span style="color: #0000BB; font-weight: bold;">other_columns</span></span></span>
<span><span class="co">#&gt; │               └── <span style="color: #0000BB; font-weight: bold;">0</span></span></span>
<span><span class="co">#&gt; │                   ├── OBJECT</span></span>
<span><span class="co">#&gt; │                   └── basic_columns.h5</span></span>
<span><span class="co">#&gt; ├── <span style="color: #0000BB; font-weight: bold;">other_data</span></span></span>
<span><span class="co">#&gt; │   ├── OBJECT</span></span>
<span><span class="co">#&gt; │   └── <span style="color: #BB0000; font-weight: bold;">list_contents.json.gz</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #0000BB; font-weight: bold;">row_data</span></span></span>
<span><span class="co">#&gt; │   ├── OBJECT</span></span>
<span><span class="co">#&gt; │   ├── basic_columns.h5</span></span>
<span><span class="co">#&gt; │   └── <span style="color: #0000BB; font-weight: bold;">other_annotations</span></span></span>
<span><span class="co">#&gt; │       ├── OBJECT</span></span>
<span><span class="co">#&gt; │       └── <span style="color: #BB0000; font-weight: bold;">list_contents.json.gz</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #0000BB; font-weight: bold;">rowgeometries</span></span></span>
<span><span class="co">#&gt; │   ├── <span style="color: #0000BB; font-weight: bold;">0</span></span></span>
<span><span class="co">#&gt; │   │   ├── OBJECT</span></span>
<span><span class="co">#&gt; │   │   └── map.parquet</span></span>
<span><span class="co">#&gt; │   └── names.json</span></span>
<span><span class="co">#&gt; └── <span style="color: #0000BB; font-weight: bold;">spatial_graphs</span></span></span>
<span><span class="co">#&gt;     ├── <span style="color: #0000BB; font-weight: bold;">0</span></span></span>
<span><span class="co">#&gt;     │   └── <span style="color: #0000BB; font-weight: bold;">col</span></span></span>
<span><span class="co">#&gt;     │       ├── <span style="color: #0000BB; font-weight: bold;">0</span></span></span>
<span><span class="co">#&gt;     │       │   ├── OBJECT</span></span>
<span><span class="co">#&gt;     │       │   ├── matrix.h5</span></span>
<span><span class="co">#&gt;     │       │   └── <span style="color: #0000BB; font-weight: bold;">method</span></span></span>
<span><span class="co">#&gt;     │       │       ├── OBJECT</span></span>
<span><span class="co">#&gt;     │       │       └── <span style="color: #BB0000; font-weight: bold;">list_contents.json.gz</span></span></span>
<span><span class="co">#&gt;     │       └── names.json</span></span>
<span><span class="co">#&gt;     └── names.json</span></span></code></pre></div>
<p>Here we see the original image is copied to the directory where we
save this SFE object. Geometries are saved as GeoParquet files, as
Apache Arrow is designed to be efficient and interoperable between
languages. It also might open the way to out of memory geometric
operations. However, note that as of writing, the GeoParquet format is
still under development and may change.</p>
<p>Now suppose we move the save directory, just like when sharing data
along with all the spatial analyses and transformations.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fsave2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fp</span>, <span class="st">"meow"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.rename</a></span><span class="op">(</span><span class="va">fsave</span>, <span class="va">fsave2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/alabaster.base/man/readObject.html" class="external-link">readObject</a></span><span class="op">(</span><span class="va">fsave2</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading SpatialExperiment</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading colgeometries</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading rowgeometries</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading spatial graphs</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading localResults</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotImage.html" class="external-link">plotImage</a></span><span class="op">(</span><span class="va">sfe2</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, channel <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">1</span>, normalize_channels <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Overview_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>The image still works. And the spatial analysis results are read as
well.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colFeatureData.html" class="external-link">colFeatureData</a></span><span class="op">(</span><span class="va">sfe2</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 10 rows and 2 columns</span></span>
<span><span class="co">#&gt;                            moran_sample01 K_sample01</span></span>
<span><span class="co">#&gt;                                 &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; transcript_counts                0.385497    24.5122</span></span>
<span><span class="co">#&gt; control_probe_counts                   NA         NA</span></span>
<span><span class="co">#&gt; control_codeword_counts                NA         NA</span></span>
<span><span class="co">#&gt; unassigned_codeword_counts             NA         NA</span></span>
<span><span class="co">#&gt; deprecated_codeword_counts             NA         NA</span></span>
<span><span class="co">#&gt; total_counts                           NA         NA</span></span>
<span><span class="co">#&gt; cell_area                        0.243229    17.2938</span></span>
<span><span class="co">#&gt; nucleus_area                           NA         NA</span></span>
<span><span class="co">#&gt; sample_id                              NA         NA</span></span>
<span><span class="co">#&gt; sizeFactor                             NA         NA</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R Under development (unstable) (2024-12-03 r87418)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8          LC_NUMERIC=C             </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=C.UTF-8           LC_COLLATE=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=C.UTF-8       LC_MESSAGES=C.UTF-8      </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8          LC_NAME=C.UTF-8          </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C.UTF-8        LC_TELEPHONE=C.UTF-8     </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=C.UTF-8    LC_IDENTIFICATION=C.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] spdep_1.3-8                    spData_2.3.3                  </span></span>
<span><span class="co">#&gt;  [3] fs_1.6.5                       sf_1.0-19                     </span></span>
<span><span class="co">#&gt;  [5] scater_1.35.0                  ggplot2_3.5.1                 </span></span>
<span><span class="co">#&gt;  [7] scuttle_1.17.0                 SingleCellExperiment_1.29.1   </span></span>
<span><span class="co">#&gt;  [9] SummarizedExperiment_1.37.0    Biobase_2.67.0                </span></span>
<span><span class="co">#&gt; [11] GenomicRanges_1.59.1           GenomeInfoDb_1.43.2           </span></span>
<span><span class="co">#&gt; [13] IRanges_2.41.1                 S4Vectors_0.45.2              </span></span>
<span><span class="co">#&gt; [15] BiocGenerics_0.53.3            generics_0.1.3                </span></span>
<span><span class="co">#&gt; [17] MatrixGenerics_1.19.0          matrixStats_1.4.1             </span></span>
<span><span class="co">#&gt; [19] SFEData_1.9.0                  Voyager_1.9.1                 </span></span>
<span><span class="co">#&gt; [21] alabaster.sfe_0.99.1           alabaster.base_1.7.2          </span></span>
<span><span class="co">#&gt; [23] SpatialFeatureExperiment_1.9.4 BiocStyle_2.35.0              </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] bitops_1.0-9              filelock_1.0.3           </span></span>
<span><span class="co">#&gt;   [3] tibble_3.2.1              R.oo_1.27.0              </span></span>
<span><span class="co">#&gt;   [5] lifecycle_1.0.4           edgeR_4.5.1              </span></span>
<span><span class="co">#&gt;   [7] lattice_0.22-6            magrittr_2.0.3           </span></span>
<span><span class="co">#&gt;   [9] limma_3.63.2              sass_0.4.9               </span></span>
<span><span class="co">#&gt;  [11] rmarkdown_2.29            jquerylib_0.1.4          </span></span>
<span><span class="co">#&gt;  [13] yaml_2.3.10               sp_2.1-4                 </span></span>
<span><span class="co">#&gt;  [15] RColorBrewer_1.1-3        DBI_1.2.3                </span></span>
<span><span class="co">#&gt;  [17] abind_1.4-8               zlibbioc_1.53.0          </span></span>
<span><span class="co">#&gt;  [19] purrr_1.0.2               R.utils_2.12.3           </span></span>
<span><span class="co">#&gt;  [21] RCurl_1.98-1.16           rappdirs_0.3.3           </span></span>
<span><span class="co">#&gt;  [23] GenomeInfoDbData_1.2.13   ggrepel_0.9.6            </span></span>
<span><span class="co">#&gt;  [25] irlba_2.3.5.1             alabaster.sce_1.7.0      </span></span>
<span><span class="co">#&gt;  [27] terra_1.7-83              units_0.8-5              </span></span>
<span><span class="co">#&gt;  [29] RSpectra_0.16-2           dqrng_0.4.1              </span></span>
<span><span class="co">#&gt;  [31] pkgdown_2.1.1             DelayedMatrixStats_1.29.0</span></span>
<span><span class="co">#&gt;  [33] codetools_0.2-20          DropletUtils_1.27.0      </span></span>
<span><span class="co">#&gt;  [35] DelayedArray_0.33.3       xml2_1.3.6               </span></span>
<span><span class="co">#&gt;  [37] tidyselect_1.2.1          UCSC.utils_1.3.0         </span></span>
<span><span class="co">#&gt;  [39] memuse_4.2-3              farver_2.1.2             </span></span>
<span><span class="co">#&gt;  [41] viridis_0.6.5             ScaledMatrix_1.15.0      </span></span>
<span><span class="co">#&gt;  [43] BiocFileCache_2.15.0      jsonlite_1.8.9           </span></span>
<span><span class="co">#&gt;  [45] BiocNeighbors_2.1.1       e1071_1.7-16             </span></span>
<span><span class="co">#&gt;  [47] systemfonts_1.1.0         tools_4.5.0              </span></span>
<span><span class="co">#&gt;  [49] ggnewscale_0.5.0          ragg_1.3.3               </span></span>
<span><span class="co">#&gt;  [51] sfarrow_0.4.1             Rcpp_1.0.13-1            </span></span>
<span><span class="co">#&gt;  [53] glue_1.8.0                gridExtra_2.3            </span></span>
<span><span class="co">#&gt;  [55] SparseArray_1.7.2         xfun_0.49                </span></span>
<span><span class="co">#&gt;  [57] EBImage_4.49.0            dplyr_1.1.4              </span></span>
<span><span class="co">#&gt;  [59] HDF5Array_1.35.2          alabaster.spatial_1.7.1  </span></span>
<span><span class="co">#&gt;  [61] withr_3.0.2               BiocManager_1.30.25      </span></span>
<span><span class="co">#&gt;  [63] fastmap_1.2.0             boot_1.3-31              </span></span>
<span><span class="co">#&gt;  [65] rhdf5filters_1.19.0       bluster_1.17.0           </span></span>
<span><span class="co">#&gt;  [67] fansi_1.0.6               rsvd_1.0.5               </span></span>
<span><span class="co">#&gt;  [69] digest_0.6.37             mime_0.12                </span></span>
<span><span class="co">#&gt;  [71] R6_2.5.1                  textshaping_0.4.0        </span></span>
<span><span class="co">#&gt;  [73] colorspace_2.1-1          wk_0.9.4                 </span></span>
<span><span class="co">#&gt;  [75] RBioFormats_1.7.0         jpeg_0.1-10              </span></span>
<span><span class="co">#&gt;  [77] RSQLite_2.3.9             R.methodsS3_1.8.2        </span></span>
<span><span class="co">#&gt;  [79] utf8_1.2.4                data.table_1.16.2        </span></span>
<span><span class="co">#&gt;  [81] class_7.3-22              httr_1.4.7               </span></span>
<span><span class="co">#&gt;  [83] htmlwidgets_1.6.4         S4Arrays_1.7.1           </span></span>
<span><span class="co">#&gt;  [85] pkgconfig_2.0.3           scico_1.5.0              </span></span>
<span><span class="co">#&gt;  [87] rJava_1.0-11              gtable_0.3.6             </span></span>
<span><span class="co">#&gt;  [89] blob_1.2.4                XVector_0.47.0           </span></span>
<span><span class="co">#&gt;  [91] htmltools_0.5.8.1         bookdown_0.41            </span></span>
<span><span class="co">#&gt;  [93] fftwtools_0.9-11          scales_1.3.0             </span></span>
<span><span class="co">#&gt;  [95] alabaster.matrix_1.7.3    png_0.1-8                </span></span>
<span><span class="co">#&gt;  [97] SpatialExperiment_1.17.0  knitr_1.49               </span></span>
<span><span class="co">#&gt;  [99] rjson_0.2.23              curl_6.0.1               </span></span>
<span><span class="co">#&gt; [101] proxy_0.4-27              cachem_1.1.0             </span></span>
<span><span class="co">#&gt; [103] rhdf5_2.51.0              BiocVersion_3.21.1       </span></span>
<span><span class="co">#&gt; [105] KernSmooth_2.23-24        vipor_0.4.7              </span></span>
<span><span class="co">#&gt; [107] parallel_4.5.0            arrow_17.0.0.1           </span></span>
<span><span class="co">#&gt; [109] AnnotationDbi_1.69.0      desc_1.4.3               </span></span>
<span><span class="co">#&gt; [111] s2_1.1.7                  pillar_1.9.0             </span></span>
<span><span class="co">#&gt; [113] grid_4.5.0                alabaster.schemas_1.7.0  </span></span>
<span><span class="co">#&gt; [115] vctrs_0.6.5               BiocSingular_1.23.0      </span></span>
<span><span class="co">#&gt; [117] dbplyr_2.5.0              beachmat_2.23.2          </span></span>
<span><span class="co">#&gt; [119] sfheaders_0.4.4           cluster_2.1.6            </span></span>
<span><span class="co">#&gt; [121] beeswarm_0.4.0            evaluate_1.0.1           </span></span>
<span><span class="co">#&gt; [123] zeallot_0.1.0             magick_2.8.5             </span></span>
<span><span class="co">#&gt; [125] cli_3.6.3                 locfit_1.5-9.10          </span></span>
<span><span class="co">#&gt; [127] compiler_4.5.0            rlang_1.1.4              </span></span>
<span><span class="co">#&gt; [129] crayon_1.5.3              labeling_0.4.3           </span></span>
<span><span class="co">#&gt; [131] classInt_0.4-10           ggbeeswarm_0.7.2         </span></span>
<span><span class="co">#&gt; [133] viridisLite_0.4.2         alabaster.se_1.7.0       </span></span>
<span><span class="co">#&gt; [135] deldir_2.0-4              BiocParallel_1.41.0      </span></span>
<span><span class="co">#&gt; [137] assertthat_0.2.1          munsell_0.5.1            </span></span>
<span><span class="co">#&gt; [139] Biostrings_2.75.1         tiff_0.1-12              </span></span>
<span><span class="co">#&gt; [141] Matrix_1.7-1              ExperimentHub_2.15.0     </span></span>
<span><span class="co">#&gt; [143] patchwork_1.3.0           sparseMatrixStats_1.19.0 </span></span>
<span><span class="co">#&gt; [145] bit64_4.5.2               Rhdf5lib_1.29.0          </span></span>
<span><span class="co">#&gt; [147] KEGGREST_1.47.0           statmod_1.5.0            </span></span>
<span><span class="co">#&gt; [149] alabaster.ranges_1.7.0    AnnotationHub_3.15.0     </span></span>
<span><span class="co">#&gt; [151] igraph_2.1.1              memoise_2.0.1            </span></span>
<span><span class="co">#&gt; [153] bslib_0.8.0               bit_4.5.0.1</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
